dali/image/jpeg2k.cc

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/image/jpeg2k.h"
+#include "dali/core/span.h"
+
+namespace dali {
+
+using block_type_t = std::array<uint8_t, 4>;
+
+namespace {
+
+constexpr block_type_t jp2_sig_type = {'j', 'P', ' ', ' '};
+
+constexpr block_type_t jp2_format_type = {'f', 't', 'y', 'p'};
+
+constexpr block_type_t jp2_header_type = {'j', 'p', '2', 'h'};
+
+constexpr block_type_t jp2_im_header_type = {'i', 'h', 'd', 'r'};
+
+constexpr int kBlockHdrSize = 8;
+
+template <typename T, int B = sizeof(T),
+          bool is_int = std::is_integral<T>::value,
+          bool is_unsigned = std::is_unsigned<T>::value>
+std::enable_if_t<is_int && is_unsigned, T> read_uint(const uint8_t *data) {
+  // We need to convert from big-endian.
+  std::array<uint8_t, B> swapped;
+  for (int i = 0; i < B; ++i)
+    swapped[B - 1 - i] = data[i];
+  T result;
+  memcpy(&result, swapped.data(), B);
+  return result;
+}
+
+inline uint32_t read_block_size(const uint8_t *data) {
+  // Size is the first 4-byte chunk of a block.
+  return read_uint<uint32_t>(data);
+}
+
+bool validate_block_type(const uint8_t *block_ptr, const block_type_t &type) {
+  // Block type is its second 4-byte chunk.
+  return span<const uint8_t, 4>(block_ptr + 4) == make_cspan(type);
+}
+
+uint32_t progress_block(span<const uint8_t> data, uint32_t index, const block_type_t &type) {
```

Comment:
Does this function advance the index to point to next block? Then:
```suggestion
uint32_t skip_block(span<const uint8_t> data, uint32_t index, const block_type_t &type) {
```
or
```suggestion
uint32_t advance_one_block(span<const uint8_t> data, uint32_t index, const block_type_t &type) {
```
?

Reply:
changed to `advance_one_block`